core/src/main/java/org/elasticsearch/action/support/replication/ReplicatedOperation.java

```diff
@@ -0,0 +1,334 @@
+        } else {
+            return null;
+        }
+    }
+
+    protected List<ShardRouting> getShards(ShardId shardId, ClusterState state) {
+        // can be null if the index is deleted / closed on us..
+        // nocommit: I'm not sure this is the right solution here. Deletion is easy (it doesn't matter), closing an index
+        // during replication may result in loss of acked writes. Solving this will required a bigger effort around closed
+        // indices.
+        final IndexShardRoutingTable shardRoutingTable = state.getRoutingTable().shardRoutingTableOrNull(shardId);
+        List<ShardRouting> shards = shardRoutingTable == null ? Collections.emptyList() : shardRoutingTable.shards();
+        return shards;
+    }
+
+    private void decPendingAndFinishIfNeeded() {
+        if (pendingShards.decrementAndGet() <= 0) {
+            finish();
+        }
+    }
+
+    private void finish() {
+        if (finished.compareAndSet(false, true)) {
+            final ReplicationResponse.ShardInfo.Failure[] failuresArray;
+            if (shardReplicaFailures.isEmpty()) {
+                failuresArray = ReplicationResponse.EMPTY;
+            } else {
+                failuresArray = new ReplicationResponse.ShardInfo.Failure[shardReplicaFailures.size()];
+                shardReplicaFailures.toArray(failuresArray);
+            }
+            finalResponse.setShardInfo(new ReplicationResponse.ShardInfo(
+                            totalShards.get(),
+                            successfulShards.get(),
+                            failuresArray
+                    )
+            );
+            finalResponseListener.onResponse(finalResponse);
+        }
+    }
+
+    private void finishAsFailed(Throwable throwable) {
+        if (finished.compareAndSet(false, true)) {
+            finalResponseListener.onFailure(throwable);
+        }
+    }
+
+
+    /**
+     * Should an exception be ignored when the operation is performed on the replica.
+     */
+    public static boolean ignoreReplicaException(Throwable e) {
+        if (TransportActions.isShardNotAvailableException(e)) {
+            return true;
+        }
+        // on version conflict or document missing, it means
+        // that a new change has crept into the replica, and it's fine
+        if (isConflictException(e)) {
+            return true;
+        }
+        return false;
+    }
+
+    public static boolean isConflictException(Throwable e) {
+        Throwable cause = ExceptionsHelper.unwrapCause(e);
+        // on version conflict or document missing, it means
+        // that a new change has crept into the replica, and it's fine
+        if (cause instanceof VersionConflictEngineException) {
+            return true;
+        }
+        return false;
+    }
+
+
+    interface Primary<Request extends ReplicationRequest<Request>, ReplicaRequest extends ReplicationRequest<ReplicaRequest>,
```

Comment:
> Can you explain what you were thinking by calling the primary proxy just Primary but the proxy to the replicas ReplicaProxy?

My original thinking was this this runs on the primary and has some arbitrary proxy to remote replicas. As I said, I think the latter is bogus (this code doesn't care about location and remote versus local). As far as this is concerned the interface is the thing.

> And the implementing class is called PrimaryShardReference

The naming in  TransportReplicationAction is indeed based on it being a reference to an actual shard. 

 If you recall I had IndexShardReference before here but people didn't like pointing at that this code does know nor care what's beneath the interface (which is in itself a reference-y word )